internal/dbutil/sqliteutil/query.go

```diff
@@ -0,0 +1,186 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package sqliteutil
+
+import (
+	"database/sql"
+	"regexp"
+	"strings"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/internal/dbutil/dbschema"
+)
+
+// QuerySchema loads the schema from sqlite database.
+func QuerySchema(db dbschema.Queryer) (*dbschema.Schema, error) {
+	schema := &dbschema.Schema{}
+
+	type Definition struct {
+		name string
+		sql  string
+	}
+
+	tableDefinitions := make([]*Definition, 0)
+	indexDefinitions := make([]*Definition, 0)
+
+	// find tables and indexes
+	err := func() error {
+		rows, err := db.Query(`
+			SELECT name, type, sql FROM sqlite_master WHERE sql NOT NULL AND name NOT LIKE 'sqlite_%'
+		`)
+		if err != nil {
+			return err
+		}
+		defer func() { err = errs.Combine(err, rows.Close()) }()
+
+		for rows.Next() {
+			var defName, defType, defSQL string
+			err := rows.Scan(&defName, &defType, &defSQL)
+			if err != nil {
+				return err
+			}
+			if defType == "table" {
+				tableDefinitions = append(tableDefinitions, &Definition{name: defName, sql: defSQL})
+			} else if defType == "index" {
+				indexDefinitions = append(indexDefinitions, &Definition{name: defName, sql: defSQL})
+			}
+		}
+
+		return rows.Err()
+	}()
+	if err != nil {
+		return nil, err
+	}
+
+	// discover tables
+	err = func() (err error) {
+		for _, definition := range tableDefinitions {
+			table := schema.EnsureTable(definition.name)
+
+			tableRows, err := db.Query(`PRAGMA table_info(` + definition.name + `)`)
+			if err != nil {
+				return err
+			}
+			defer func() { err = errs.Combine(err, tableRows.Close()) }()
+
+			for tableRows.Next() {
+				var defaultValue sql.NullString
+				var index, name, columnType string
+				var pk int
+				var notNull bool
+				err := tableRows.Scan(&index, &name, &columnType, &notNull, &defaultValue, &pk)
+				if err != nil {
+					return err
+				}
+
+				column := &dbschema.Column{
+					Name:       name,
+					Type:       columnType,
+					IsNullable: !notNull && pk == 0,
+				}
+				table.AddColumn(column)
+				if pk > 0 {
+					if table.PrimaryKey == nil {
+						table.PrimaryKey = make([]string, 0)
+					}
+					table.PrimaryKey = append(table.PrimaryKey, name)
+				}
+
+			}
+
+			matches := rxUnique.FindAllStringSubmatch(definition.sql, -1)
+			for _, match := range matches {
+				// TODO feel this can be done easier
+				var columns []string
+				for _, name := range strings.Split(match[1], ",") {
```

Comment:
just clarification question- what's at `match[0]`? I also feel like there might be more efficient way to do this but not sure what it is.

Reply:
`match[0]` is the full regular expression match... `match[1]` is the first group match.

Performance doesn't matter that much here since 1. it's only for testing and it's iterating the table schema, which at most is ~10 tables.